source/io/module_eggShell.js

```diff
@@ -361,29 +386,133 @@
+            }
         };
 
         Module.eggShell.executeSlicesUntilWait = publicAPI.eggShell.executeSlicesUntilWait = function (slices) {
+            // Returns an ExecSlicesResult:
             // returns < 0 if should be called again ASAP, 0 if nothing to run, or positive value N if okay
             // to delay up to N milliseconds before calling again
             return EggShell_ExecuteSlices(v_userShell, slices);
         };
 
+        // performanceNow is a small polyfill for the browser Performance.now api to use in node.js
+        var performanceNow = (function () {
+            var result;
+
+            var nodeStartTimeMS;
+            var nodeTimeMS = function () {
+                var hrtime = process.hrtime();
+                return (hrtime[0] * 1000) + (hrtime[1] / 1e6);
+            };
+
+            if (typeof performance !== 'undefined') {
+                result = function () {
+                    return performance.now();
+                };
+            } else if (typeof process !== 'undefined') {
+                nodeStartTimeMS = nodeTimeMS();
+                result = function () {
+                    return nodeTimeMS() - nodeStartTimeMS;
+                };
+            } else {
+                result = function () {
+                    throw new Error('Platform unsupported, require either performance.now or process.hrtime timing api');
+                };
+            }
+            return result;
+        }());
+
+        // Pumps vireo asynchronously until the currently loaded via has finished all clumps
+        // Runs synchronously for a maximum of 4ms at a time to cooperate with browser and node.js execution environments
+        // A good starting point for most vireo uses but can be copied and modified as needed
+        // If a callback (stdout, stderr) is provided, it will be run asynchronously to completion
+        Module.eggShell.executeSlicesUntilClumpsFinished = publicAPI.eggShell.executeSlicesUntilClumpsFinished = function (callback) {
+            var MAXIMUM_VIREO_EXECUTION_TIME_MS = 4;
+
+            var timerToken;
+            var origExecuteSlicesWakeUpCallback = Module.eggShell.executeSlicesWakeUpCallback;
+
+            var vireoFinished = function () {
+                Module.eggShell.executeSlicesWakeUpCallback = origExecuteSlicesWakeUpCallback;
+
+                if (typeof callback === 'function') {
+                    callback();
+                }
+            };
+
+            var runExecuteSlicesAsync = function () {
+                var execSlicesResult, elapsedTime;
+                var startTime = performanceNow();
+
+                do {
+                    execSlicesResult = Module.eggShell.executeSlicesUntilWait(100000);
+                    elapsedTime = performanceNow() - startTime;
+
+                    if (execSlicesResult >= 0) {
+                        break;
+                    }
+                } while (elapsedTime < MAXIMUM_VIREO_EXECUTION_TIME_MS);
```

Comment:
The C entry point for ExecuteSlices has a maximum execute time parameter which is not exposed to the JS wrapper.   Maybe instead of duplicating that functionality here, we should just expose the parameter to JS.

Reply:
Oh yea forgot about that. And if I understand it right the current hardcoded value in ExecutionSlices is 20 ticks and from TickCount ticks are microseconds so it's effectively unused as is.

Yea I'll try exposing it and see if it behaves as expected. I'll leave the review open for now for other feedback since that will be a pretty small scope change on top of this.

Reply:
Worked well! Exposed to the JS side and deleted unnecessary code.

Reply:
Hurray!